### DISCUSSION TITLE
Interpret wildcards in paths as partitions?

### DISCUSSION THREAD

**Discussion by liquidcarbon** [duckdb/duckdb#8487] [Q&A]
My `table` consists of ~200K parquet files with two-level partitioning.

The query `FROM '/data/v1.1/partitionA=XYZ/partitionB=123/table.parquet'` returns the result immediately, as expected.
The query `FROM '/data/v1.1/*/*/table.parquet' WHERE partitionA='XYZ' and partitionB=123` takes >20 seconds.  Also expected if DuckDB scanned the whole folder structure.  But I was hoping it could somehow recognize the `/*/*/` in path as partition levels.  Is there a way to help DuckDB to be clever like that around wildcards?

I'm building an API to query by multiple specific partitions, and it seems easier to define the source table once and filter partitions using WHERE IN (`WHERE partitionA IN (x, y)`)  than separate FROMs:

```
    FROM (
        SELECT * FROM
        '/data/v1.1/partitionA=XYZ/partitionB=123/table.parquet'
        UNION
        SELECT * FROM
         '/data/v1.1/partitionA=XYZ/partitionB=456/table.parquet'
    )
```
Further, if I went the UNION route and asked for a partition that doesn't exist, it errors out:
`IO Error: No files found that match the pattern "/data/v1.1/partitionA=XYZ/partitionB=does_not_exist/table.parquet"`

**Comment by Tishj** [comment]
I believe this fixes that issue
https://github.com/duckdb/duckdb/pull/8428

**Reply by liquidcarbon** [reply]
neat! What's your best guess for when the `HIVE_FILTER` option arrives?

**Comment by liquidcarbon** [comment]
EDIT: this also suffers from a hard error in case when even one of the specified files are missing

`query did not run: IO Error: No files found that match the pattern "/data/v1.1/.../.../....parquet"`

```
-- use list parameter to read 3 parquet files and treat them as a single table
SELECT * FROM read_parquet(['file1.parquet', 'file2.parquet', 'file3.parquet']);
```

This offers a reasonable workaround.  Given a dataframe that contains partitions of interest, one can assemble a list of files to read:
```
def format_from(df: pd.DataFrame, table: str):
    """Format FROM clause as a list of parquet files.
    
    Input: DataFrame with columns `A` and `B`
    """
    FROM = [
        f"{DATA_ROOT}/partitionA={a}/partitionB={b}/{table}.parquet"
        for a, b
        in zip(df.A, df.B)
    ]
    return FROM
```

Then write query:

```
q = f"""
SELECT ...
FROM READ_PARQUET({format_from(df, 'mytable')})
"""
```


Requires running more than one query, but that's ok.

**Comment by liquidcarbon** [comment]
There seems to be an improvement in 0.9.2, but the most direct way remains quite slow.
The likely fix https://github.com/duckdb/duckdb/pull/8428 remains closed at this time
